Synthetic case generation workflow with sample medication_management case set

aaryanxshah opened the pull request.

andhines commented:
Please investigate these warnings from the test runs:
```
================================================================ warnings summary ================================================================
tests/evaluations/synthetic_unit_cases/test_eval_report_generator.py::test_script_generates_csv
tests/evaluations/synthetic_unit_cases/test_eval_report_generator.py::test_length_mismatch_raises
  <frozen runpy>:128: RuntimeWarning: 'evaluations.cases.synthetic_unit_cases.eval_report_generator' found in sys.modules after import of package 'evaluations.cases.synthetic_unit_cases', but prior to execution of 'evaluations.cases.synthetic_unit_cases.eval_report_generator'; this may result in unpredictable behaviour

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================== 697 passed, 2 warnings in 0.65s =========================================================
```

aaryanxshah commented on evaluations/cases/synthetic_unit_cases/grader_generator.py:
```diff
@@ -0,0 +1,106 @@
+        except json.JSONDecodeError:
+            print("LLM produced invalid JSON — saving raw text for inspection.")
+            self.output_path.write_text(cleaned)
+            sys.exit(1)
```
Would love to discuss better ways to handle this but the rest of the script for extracting the rationale and given score ends up failing if json isn't well-formed–felt preferable to terminate to figure out the issue to change prompts/script.

dbajet commented on hyperscribe/libraries/constants.py:
```diff
@@ -14,14 +14,15 @@ class Constants:
     MEMORY_LOG_LABEL = "main"
     OPENAI_CHAT_AUDIO = "gpt-4o-audio-preview"  # LLM model used for speech to text
-    OPENAI_CHAT_TEXT = "gpt-4o"  # LLM model used for text completion
+    OPENAI_CHAT_TEXT = "o3"  # LLM model used for text completion
```
Using `o3` by default will terribly slow `hyperscribe`.

aaryanxshah replied:
Yes, this has been changed. Hyperscribe using 4o, case generation using o3.

dbajet replied:
Please, let the author of a comment to mark it as resolved, otherwise, it will be difficult for the next review to remember what were  the issues.

dbajet commented on hyperscribe/libraries/constants.py:
```diff
@@ -14,14 +14,15 @@ class Constants:
     FAUX_NOTE_UUID = "_NoteUuid"
     FAUX_PATIENT_UUID = "_PatientUuid"
-    FAUX_PROVIDER_UUID = "_ProviderUuid"
+    FAUX_PROVIDER_UUID = "4150cd20de8a470aa570a852859ac87e"
```
would `11111111111111111111111111111111` be ok?
this way it is easy to see that's intentional to have a fake uuid

dbajet replied:
I don't understand the need of a "Canvas-compatible" uuid, and `11111111111111111111111111111111` (or `a000000000000001000000000000001a`) should also be compatible.

If you mean that `4150cd20de8a470aa570a852859ac87e` is an existing one, then it would exist only on one instance of Canvas.

Could you elaborate on the need  of a "Canvas-compatible" uuid?

dbajet commented on hyperscribe/libraries/constants.py:
```diff
@@ -65,6 +66,8 @@ class Constants:
     SECRET_AUDIO_KEY = "KeyAudioLLM"
     SECRET_AUDIO_VENDOR = "VendorAudioLLM"
     SECRET_AUDIT_LLM = "AuditLLMDecisions"
+    SECRET_TEXT_MODEL = "TEXT_MODEL"
```
if the model is now part of the settings/secret, why not using it in `hyperscribe/libraries/helper.py::chatter`?

dbajet replied:
All the secrets are following the camel case notation, except `TEXT_MODEL` and `TEXT_TEMPERATURE`...

dbajet commented on hyperscribe/libraries/helper.py:
```diff
@@ -46,11 +46,48 @@ def icd10_strip_dot(cls, code: str) -> str:
+            client.model = requested
+            return client
+
+        # default to OpenAI “gpt-4o” model
```
The comment is incorrect as in constant, we can see `OPENAI_CHAT_TEXT = "o3"`

aaryanxshah commented on case_builder.py:
```diff
@@ -22,9 +30,31 @@ def run(cls, arguments: list[str]) -> None:
+        #edited control flow and pulled this out of elif to make sure it runs if in the argument, not dependent now on other builder files. 
+        #because of this, we also don't need to have --rubric be parsed in any other builder files.
+        '''
+        if "--rubric" in arguments:
```
No longer valid and has been removed.

aaryanxshah commented on evaluations/case_builders/builder_from_chart_transcript.py:
```diff
@@ -0,0 +1,83 @@
+from hyperscribe.structures.line import Line
+
+
+class BuilderFromChartTranscript(BuilderBase):
```
this will be resolved after this pr with refactoring to work with Postgres interface, as discussed durign meeting today.

dbajet commented on tests/test_case_builder.py:
```diff
@@ -37,43 +39,36 @@ def reset_mocks():
 
-        assert capsys.readouterr().out == exp_out
+    for (arguments, exp_out, call_delete, call_transcript, call_tuning, call_mp3, 
+         call_audit, call_summarize, call_chart_transcript, call_direct_split, call_direct_full,) in tests:
```
having one item per line makes comparisons and count easier

dbajet reviewed (changes requested):
Thank you @aaryanxshah for the efforts made to follow a lot of my comments and code changes requests.

There are still some changes to consider, and, in general:
 + the name of the classes should be the same than the files (besides the snake or camel cases formats)
 + abbreviations in computer code are generally discouraged unless they are widely recognized or follow established conventions (e.g. HTTP)
 + it seems that the user prompts have not been reviewed after changing the LLM discussion to use the the JSON markdown blocks in `evaluations/case_builders/synthetic_json_helper.py::generate_json`
 + the JSON Schema should be added to the first user prompt
 + the standalone methods should be avoided, especially when having the same name (`main`), otherwise, it prevent importing together these files (without side effect)

An important change, but not difficult would be to refactor `evaluations/case_builders/synthetic_json_helper.py` as explained.

For the tests (applicable to almost all tests of the PR):
 + it is important to keep the order of the tests and the functions (most added to tests to existing files are at the end, e.g `test_list_s3_objects_with_continuation`)
 + all methods are covered but not tested (e.g. `test_grader.py`)
 + using `monkeypatch` should be limited to cases when `patch` is impossible/difficult
 + when mocking, using `side_effect` over `return_value` gives more control and flexibility
 + checking the mock calls with `mock_calls` rather than the `assert_call...` methods ensures tight tests
 + no prompt or JSON Schema is part of the tests


(you can resolve all comments that are just thanking you for the underlined changes :wink: )

dbajet commented on tests/evaluations/case_builders/test_synthetic_profile_generator.py:
```diff
@@ -0,0 +1,173 @@
+    #system prompt validation
+    system_prompt = kwargs["system_prompt"]
+    assert isinstance(system_prompt, list)
+    assert any("clinical‑informatics expert" in line for line in system_prompt)
```
This is a good attempt to assert some parts of the text are there.
But using the MD5 (or any other hash) is simpler and sure (and less cumbersome than checking the full text)

dbajet commented on tests/evaluations/case_builders/test_synthetic_profile_generator.py:
```diff
@@ -0,0 +1,173 @@
+    assert tested.all_profiles == {}
+
+def test_extract_initial_fragment(vendor_key: VendorKey):
+    tested = SyntheticProfileGenerator(vendor_key, "out.json")
```
`"out.json"`: a `Path` is expected.

dbajet commented on tests/evaluations/case_builders/test_synthetic_profile_generator.py:
```diff
@@ -0,0 +1,173 @@
+@patch.object(SyntheticProfileGenerator, "_save_combined")
+@patch.object(SyntheticProfileGenerator, "generate_batch")
+def test_run(mock_generate_batch, mock_save_combined, mock_save_individuals, vendor_key: VendorKey):
+    tested = SyntheticProfileGenerator(vendor_key, "out.json")
```
`"out.json"`: a `Path` is expected.

dbajet commented on tests/evaluations/case_builders/test_synthetic_profile_generator.py:
```diff
@@ -0,0 +1,173 @@
+
+    expected_calls = [call(i, batch_size) for i in range(1, batches + 1)]
+    mock_generate_batch.assert_has_calls(expected_calls, any_order=False)
+    mock_save_combined.assert_called_once()
```
call once... but with what?

dbajet commented on tests/hyperscribe/libraries/test_helper.py:
This file is actually unchanged besides some space and empty lines.
Just revert its changes.

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
Not all methods are tested (`_random_bucket`...)

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+@patch.object(random, "uniform", return_value=1.25)
+@patch.object(random, "randint", return_value=3)
+@patch.object(random, "choice")
+def test_make_spec_deterministic(mock_choice, _randint, _uniform, tmp_path):
```
the test should be called `test__make_spec`

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+    assert spec["speaker_sequence"] == ["Clinician", "Patient", "Patient"]
+    assert spec["ratio"] == 1.25
+    assert all(m in SpecConstants.MOOD_POOL for m in spec["mood"])
+
```
mocks' calls are not checked

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+    system_lines, user_lines = tested._build_prompt("profile text", spec, dummy_schema)
+    combined_system = "\n".join(system_lines)
+
+    assert "Avoid starting with any of these previous first lines: previous first line" in combined_system
```
Use the MD5 approach to note any change

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+
+
+@patch.object(HelperSyntheticJson, "generate_json", return_value=[{"speaker": "Clinician", "text": "Sample text"}])
+def test_generate_transcript__success(mock_generate_json, tmp_path):
```
the test should be named `test_generate_transcript_for_profile__success`

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+
+
+@patch.object(HelperSyntheticJson, "generate_json", side_effect=ValueError("Invalid JSON"))
+def test_generate_transcript__bad_json_raises(mock_generate_json, tmp_path):
```
the test should be named `test_generate_transcript_for_profile__bad_json_raises`

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+    calls = [call(*args, **kwargs)]
+    assert mock_generate_json.mock_calls == calls
+    assert result_transcript == expected_transcript
+    assert result_transcript[0]["text"].lower().strip() in tested.seen_openings
```
```
assert result_transcript[0]["text"].lower().strip() in tested.seen_openings
```
should be
```
assert "sample text" in tested.seen_openings
```
simpler, and not redoing what the code does (which is an anti-pattern)

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+    args = mock_generate_json.call_args.args
+    kwargs = mock_generate_json.call_args.kwargs
+    calls = [call(*args, **kwargs)]
+    assert mock_generate_json.mock_calls == calls
```
this assert is equivalent to `assert var == var`... which is not a real test

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,134 @@
+    args = mock_generate_json.call_args.args
+    kwargs = mock_generate_json.call_args.kwargs
+    calls = [call(*args, **kwargs)]
+    assert mock_generate_json.mock_calls == calls
```
same thing, this assert is equivalent to assert var == var... which is not a real test

dbajet commented on tests/hyperscribe/libraries/test_aws_s3.py:
```diff
@@ -349,6 +349,103 @@ def rest_mocks():
+    objs = client.list_s3_objects("some/prefix")
+    assert len(objs) == 6
+    assert all(isinstance(o, AwsS3Object) for o in objs)
+    assert is_ready.mock_calls == [call()]
```
Thanks!

dbajet commented on tests/hyperscribe/libraries/test_aws_s3.py:
```diff
@@ -407,3 +504,8 @@ def reset_mocks():
     assert mock_datetime.mock_calls == []
     reset_mocks()
+
+'''
```
Comment misplaced?

dbajet commented on tests/hyperscribe/llms/test_llm_openai.py:
```diff
@@ -236,3 +238,56 @@ def reset_mocks():
+
+#status != 200 -> returns full raw JSON. 
+@patch("hyperscribe.llms.llm_openai.requests_post")
+def test_request_passes_through_non_200(requests_post):
```
Thank you for moving the tests.

While, again, they don't cover a case not previously covered, it is better to have more than not enough

dbajet commented on tests/hyperscribe/llms/test_llm_openai_o3.py:
```diff
@@ -0,0 +1,17 @@
+    result = llm.to_dict(for_log=True)
+
+    assert "modalities" not in result
+    assert result == {"other": "value"}
```
The case `"modalities"` is not present in the original dictionary could be added

dbajet commented on tests/hyperscribe/structures/test_settings.py:
```diff
@@ -172,3 +172,4 @@ def test_list_from():
     for string, expected in tests:
         result = tested.list_from(string)
         assert result == expected, f"---> {string}"
+        
```
It seems that the files is unchanged beside an extra empty line, just revert it

dbajet commented on tests/hyperscribe/structures/test_vendor_key.py:
```diff
@@ -1,11 +1,11 @@
         "api_key": str,
     }
     assert is_namedtuple(tested, fields)
+    
```
It seems that the files is unchanged beside an extra empty line, just revert it

dbajet commented on tests/test_case_builder.py:
```diff
@@ -37,43 +39,45 @@ def reset_mocks():
+         call_direct_full) in tests:
+        CaseBuilder.run(arguments)
 
+        result = capsys.readouterr().out
```
Good catch!

aaryanxshah commented on evaluations/case_builders/note_grader.py:
```diff
@@ -0,0 +1,124 @@
+            "",
+            "Each element keys:",
+            "  • rationale    : str  – short explanation",
+            "  • satisfaction : int – 0 to 100",
```
changed to include only ascii characters

aaryanxshah commented on evaluations/case_builders/synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,163 @@
+            "",
+            "Moods: " + ", ".join(spec["mood"]),
+            "External pressure: " + spec["pressure"],
+            f"Clinician persona: {spec['clinician_style']}",
```
This has been changed to all f-strings

aaryanxshah commented on tests/evaluations/case_builders/test_helper_synthetic_json.py:
```diff
@@ -0,0 +1,101 @@
+        schema={"type": "object", "properties": {"key": {"type": "string"}}, "required": ["key"]},)
+
+    assert tested == expected
+    assert not (tmp_path / "invalid_output.json").exists()
```
As a result of the removal of some of the validation logic (already handled by llm.chat), I have removed one of the tests within test_helper_synthetic_json.

dbajet reviewed (changes requested):
Thank you @aaryanxshah for the changes.

There are a few real concerns or errors:
* evaluations/case_builders/synthetic_chart_generator.py (concern)
* evaluations/case_builders/synthetic_profile_generator.py (error!)

And besides this, `Dict`and `List` could be changed to `dict` and `list`  every where.

In regards for the tests, again, thank you for the effort.
Please look at the `tests/evaluations/case_builders/test_helper_synthetic_json.py` for not using `ANY`.

Otherwise, there a few suggestions (removing useless asserts, using actual values rather than computed ones...).

But in general, my concern is the use of the same logic (same code?) in the actual code and the tests (tests/evaluations/case_builders/test_synthetic_transcript_generator.py).

And, yes, the tests will break if the code is changed (something I pointed out as a goal in the last review), but the first goal of the tests remain to check that what is done is correctly done.
So using the same code in both sides defeat this purpose.

In this specific case, since most of the duplications are related to prompts, the concern is limited though.

aaryanxshah commented:
Thank you @dbajet for the comments on this last iteration. I have updated each of the files as requested, along with a consistent md5-based approach for checking prompts in the case generation files that removes the duplication efforts in the tests that existed before. Additionally, all schemas have been mocked as mentioned before. `Dict` and `List` have been removed  and replaced with `dict` and `list` across all case generation files.

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+        tested.generate_transcript_for_profile(profile_text)
+    assert "Invalid JSON" in str(exc_info.value)
+
+    _, kwargs = mock_generate_json.call_args
```
`call_args` is related to the last call of `mock_generate_json`.
Since `return_value` is used, we don't know if there re more than one call (with `side_effect`, an error occurs when there is not enough provide values for the calls)
So a check on the count of calls should be done, and if there are more than 1, the other calls need to be checked.

aaryanxshah replied:
logic adjusted as mentioned in previous comment with mock_call == call params

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+    assert (result_directory / "transcript.json").exists()
+    assert (result_directory / "spec.json").exists()
+
+    _, kwargs = mock_generate_json.call_args
```
`call_args` is related to the last call of `mock_generate_json`.
Same consequences than above.

aaryanxshah replied:
logic adjusted as mentioned in previous comment with mock_call == call params

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+
+    assert mock_generator_class.call_args == call(vendor_key=fake_vendor_key,
+        input_path=fake_input, output_path=fake_output)
+    assert mock_instance.run.call_args == call(start_index=5, limit=10)
```
`call_args` is related to the last call of `mock_generate_json`.
Same consequences than above.

aaryanxshah replied:
adjusted to avoid call_args logic because as you mentioned, it only checks the last call. although most mocks are only called once, better to keep consistent with `assert mock_generate_json.mock_calls == [call(
        vendor_key=tested.vendor_key,
        schema=expected_schema,
        system_prompt=["System Prompt"],
        user_prompt=["User Prompt"]
    )]`

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+@patch("evaluations.case_builders.synthetic_transcript_generator.SyntheticTranscriptGenerator")
+@patch("evaluations.case_builders.synthetic_transcript_generator.HelperEvaluation.settings")
+@patch("evaluations.case_builders.synthetic_transcript_generator.argparse.ArgumentParser.parse_args")
+def test_main(mock_parse_args, mock_settings, mock_generator_class):
```
no checks on calls for `mock_parse_args`, `mock_settings`

aaryanxshah replied:
rewritten to check on mock_calls via mock_calls == [call()]

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+        "patient_style": ""})
+@patch.object(HelperSyntheticJson, "generate_json", return_value=[{"speaker": "Clinician", "text": "Content"}])
+def test_run(mock_generate_json, mock_make_spec, mock_build_prompt, mock_schema_transcript, tmp_path):
+    input_file = create_fake_profiles_file(tmp_path)
```
no checks on the calls of `mock_make_spec`, `mock_build_prompt`, `mock_schema_transcript`.

aaryanxshah replied:
rewritten to check on mock_calls

aaryanxshah commented on tests/evaluations/case_builders/test_note_grader.py:
```diff
@@ -0,0 +1,183 @@
+
+    run_calls = {}
+    def fake_run(self):
+        run_calls["inst"] = self
```
renamed as "instance"

aaryanxshah commented on tests/evaluations/case_builders/test_rubric_generator.py:
```diff
@@ -0,0 +1,125 @@
+                                  "minimum": 0, "maximum": 100},
+                    "sense":     {"type": "string", 
+                                  "description": "positive or negative direction",
+                                  "enum": [Constants.POSITIVE_VALUE, Constants.NEGATIVE_VALUE]}
```
these values have been changed to `"positive"` and '"negative"'

aaryanxshah commented on tests/evaluations/case_builders/test_rubric_generator.py:
```diff
@@ -0,0 +1,125 @@
+    assert kwargs["schema"] == expected_schema
+    assert result_system_md5 == expected_system_md5
+    assert result_user_md5 == expected_user_md5
+    
```
done, but `call(count)` instead of call() given the structure of mock_schema_batch.

dbajet replied:
I don't understand your answer: `assert mock_schema_rubric.mock_calls == [call()]` is added on line 82

aaryanxshah commented on evaluations/case_builders/synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,168 @@
+
+        first  = random.choice(["Clinician", "Patient"])
+        other  = "Patient" if first == "Clinician" else "Clinician"
+        seq    = [first] + [random.choice([first, other]) for _ in range(turn_total - 1)]
```
changed to the latter

aaryanxshah commented on evaluations/case_builders/synthetic_chart_generator.py:
```diff
@@ -0,0 +1,159 @@
+
+        properties = {
+            key: {"type": "array" if isinstance(value, list) else "string",
+                "description": _DESCRIPTIONS.get(key, "")}
```
rewritten to '_DESCRIPTIONS[key]' to allow for 'KeyError' if missing.

aaryanxshah commented on tests/evaluations/case_builders/test_synthetic_chart_generator.py:
```diff
@@ -0,0 +1,175 @@
+    properties = result_schema["properties"]
+    assert set(properties.keys()) == {"demographicStr", "customData"}
+    #a correct description string appended, as well as an empty one.
+    assert properties["demographicStr"] == {"type": "string", "description": "string describing patient demographics"}
```
thank you for this thoughtful comment–I have moved the description map into the `evaluations/constants.py` file where descriptions will be checked there, along with the KeyError raising previously mentioned.

aaryanxshah commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+    mock_schema_transcript.side_effect = lambda spec: expected_schema
+    transcript, spec = tested.generate_transcript_for_profile(profile_text)
+
+    _, kwargs = mock_generate_json.call_args
```
rewritten for consistent logic

aaryanxshah commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+        "clinician_style": "",
+        "patient_style": ""})
+@patch.object(HelperSyntheticJson, "generate_json", return_value=[{"speaker": "Clinician", "text": "Sample text"}])
+def test_generate_transcript_for_profile__success(mock_generate_json, mock_make_spec, mock_build_prompt, mock_schema_transcript, tmp_path):
```
rewritten for consistent logic

aaryanxshah commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,258 @@
+@patch.object(random, "randint", return_value=3)
+@patch.object(random, "sample") 
+@patch.object(random, "choice")
+def test__make_spec(mock_choice, mock_sample, _randint, _uniform, tmp_path):
```
rewritten and adjusted

dbajet commented on tests/evaluations/case_builders/test_rubric_generator.py:
```diff
@@ -0,0 +1,127 @@
+    with pytest.raises(SystemExit) as exc:
+        tested.generate(transcript, chart, context, output_path=output_path)
+    assert exc.value.code == 1
+    assert len(mock_generate_json.mock_calls) == 1
```
I guess this is where the [previous comment](https://github.com/canvas-medical/canvas-hyperscribe/pull/57/files?file-filters%5B%5D=.md&file-filters%5B%5D=.py&show-deleted-files=true&show-viewed-files=true#r2223658178) makes sense.

This is where having one `test_` function holding several cases through a loop is helpful (but more complex, which may be a deterrent for tests):
```python
tests = [
 (val1, val2, val3),
 (val4, val5, val6),
]
for parm1, parm2, expected in tests:
  ...
```

dbajet commented on tests/evaluations/case_builders/test_synthetic_chart_generator.py:
```diff
@@ -0,0 +1,185 @@
+    result = SyntheticChartGenerator.load_json(data_file)
+    assert result == expected
+
+from evaluations.constants import Constants
```
the import should be at the beginning of the file

dbajet commented on tests/evaluations/case_builders/test_synthetic_chart_generator.py:
```diff
@@ -0,0 +1,185 @@
+    properties = result_schema["properties"]
+    assert set(properties.keys()) == set(example_chart_known.keys())
+
+    for key, expected_desc in Constants.EXAMPLE_CHART_DESCRIPTIONS.items():
```
While you can use the constants for input in a tests, you never use constants for checking the output but their actual values.

dbajet commented on tests/evaluations/case_builders/test_synthetic_chart_generator.py:
```diff
@@ -0,0 +1,185 @@
+    tested_chart = {"bad": True}
+    tested = SyntheticChartGenerator(VendorKey("v", "k"), {}, Path("."), {})
+    result = tested.validate_chart(tested_chart)
+    assert result == False
```
`assert result == False` -> `assert result is False`

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,276 @@
+        ["short"]
+        + ["Clinician"]
+        + ["Patient"] * 2
+        + [SpecificationConstants.PRESSURE_POOL[0]]
```
As mentioned above (and sorry for missing this in the previous reviews), constants should never be used when checking outputs.

dbajet commented on tests/evaluations/case_builders/test_synthetic_transcript_generator.py:
```diff
@@ -0,0 +1,276 @@
+        call(["Clinician", "Patient"]),
+        call(["Clinician", "Patient"]),
+        call(["Clinician", "Patient"]),
+        call(SpecificationConstants.PRESSURE_POOL),
```
As mentioned above (and sorry for missing this in the previous reviews), constants should never be used when checking outputs.

dbajet reviewed (approved):
Hurray! :tada:

The code is robust and the tests comprehensive.

There are comments to consider, and minor changes remaining:
- `tests/evaluations/case_builders/test_synthetic_chart_generator.py`
- `tests/evaluations/case_builders/test_synthetic_transcript_generator.py`

So, once done, this is ready for merge.
